charmcraft/extensions/gunicorn.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2023-2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""The flask extension."""
+import abc
+from typing import Any
+
+from overrides import override
+
+from ..errors import ExtensionError
+from .extension import Extension
+
+
+class _GunicornBase(Extension):
+    """A base class for 12-factor WSGI applications."""
+
+    _ACTIONS = {
+        "flask": {
+            "rotate-secret-key": {
+                "description": "Rotate the flask secret key. Users will be forced to log in again. This might be useful if a security breach occurs."
+            }
+        },
+    }
```

Comment:
What I'm looking for here is something like:

```python
class _GunicornBase(Extension):
    """A base class for 12-factor WSGI applications."""

    _ACTIONS: dict[str, dict[str, str]]

...

class FlaskFramework(_GunicornBase):
    """Extension for 12-factor Flask applications."""

    _ACTIONS = {
        "rotate-secret-key": {
            "description": "Rotate the flask secret key. Users will be forced to log in again. This might be useful if a security breach occurs."
        }
    }
```

That way all the flask-specific stuff will be together and the django-specific stuff will be together.

Since `_WEBSERVER_OPTIONS` is generic across all gunicorn extensions we can leave it in the gunicorn abstract base.

Reply:
that's super clever and elegant and requires less indentation! updated, thank you!